src/lib/phdi-building-blocks/phdi_building_blocks/geo.py

```diff
@@ -0,0 +1,119 @@
+    """
+
+    lookup = Lookup(street=address)
+    client.send_lookup(lookup)
+
+    if lookup.result and lookup.result[0].metadata.latitude:
+        res = lookup.result[0]
+        addr = [res.delivery_line_1]
+        if res.delivery_line_2:
+            addr.append(res.delivery_line_2)
+
+        return GeocodeResult(
+            address=addr,
+            city=res.components.city_name,
+            state=res.components.state_abbreviation,
+            zipcode=res.components.zipcode,
+            county_fips=res.metadata.county_fips,
+            county_name=res.metadata.county_name,
+            lat=res.metadata.latitude,
+            lng=res.metadata.longitude,
+            precision=res.metadata.precision,
+        )
+
+
+def get_smartystreets_client(auth_id: str, auth_token: str) -> us_street.Client:
+    """
+    Build a smartystreets api client from an auth id and token
+    """
+
+    creds = StaticCredentials(auth_id, auth_token)
+
+    return (
+        ClientBuilder(creds)
+        .with_licenses(["us-standard-cloud"])
+        .build_us_street_api_client()
+    )
+
+
+def geocode_patient_address(bundle: dict, client: us_street.Client) -> dict:
+    """Given a FHIR bundle and a SmartyStreets client, geocode all patient addresses
+    in all patient resources in the bundle."""
+
+    for resource in find_patient_resources(bundle):
+        patient = resource.get("resource")
+        if "extension" not in patient:
+            patient["extension"] = []
+
+        raw_addresses = []
+        std_addresses = []
+        for address in patient.get("address", []):
+            # Generate a one-line address to pass to the geocoder
+            one_line = " ".join(address.get("line", []))
+            one_line += f" {address.get('city')}, {address.get('state')}"
+            if "postalCode" in address and address["postalCode"]:
+                one_line += f" {address['postalCode']}"
+            raw_addresses.append(one_line)
+
+            geocoded = geocode(client, one_line)
+            std_one_line = ""
+            if geocoded:
+                address["line"] = geocoded.address
+                address["city"] = geocoded.city
+                address["state"] = geocoded.state
+                address["postalCode"] = geocoded.zipcode
+                std_one_line = f"{geocoded.address} {geocoded.city}, {geocoded.state} {geocoded.zipcode}"  # noqa
+                std_addresses.append(std_one_line)
+
+                if "extension" not in address:
+                    address["extension"] = []
+
+                address["extension"].append(
+                    {
+                        "url": "http://hl7.org/fhir/StructureDefinition/geolocation",
+                        "extension": [
+                            {"url": "latitude", "valueDecimal": geocoded.lat},
+                            {"url": "longitude", "valueDecimal": geocoded.lng},
+                        ],
+                    }
+                )
```

Comment:
This may not be the place for it but I actually just spotted the error that was stopping the metrics tracking from successfully calculating during the one trial run before the demo. If the geocoder fails, then `std_addresses` doesn't get an addition, which means the length of the two arrays could get mismatched. To combat this, we need to do two things:
1. in line 88, instead of setting `std_one_line = ""`, we should have `std_one_line = one_line`.
2. we should remove line 95 and instead call `std_addresses.append(std_one_line)` right before we perform the `any_didffs` calculation.
These changes ensure the lengths of the two arrays are the same and also allow for correct stats tabulation: in the current case, if geocoding fails, we count that as a diff and so we'll get a +1 bump to our number of records that had the address changed. But this is a vacuous stat that we shouldn't get credit for, so instead appending the same raw one line if geocoding fails ensures that it won't count this as the standardization working.

Reply:
What you have said makes sense, but I would propose making these corrections in a separate PR around tracking for pipeline metrics. While refactoring I was also wondering if we could bundle this tracking into another module that can be called internally from other building blocks. Maybe more trouble than it is worth, but might be nice for developing additional transforms down the road and easily adding logging.

Reply:
Yeah that's totally fair Dan. I figured a reorganization PR isn't the place to do that. It's at least somewhat comforting to know though that we've identified the early candidate for the error-cause in metrics tracking at least. And I think eventually we probably should bundle the tracking separately into another module (perhaps putting it in the `utils.py` building block). That will likely be quite difficult because it'll involve creating a framework rather than tagging for a specific purpose, but either way, it's beyond scope for right now.